examples/apps/color_finder.py

```diff
@@ -0,0 +1,659 @@
+
+
+TODO: determine whether to keep pixel_matrix as is, or convert to numpy matrix
+'''
+
+import asyncio
+import functools
+import math
+import numpy
+import sys
+
+import cozmo
+
+from cozmo.util import degrees, radians, distance_mm, speed_mmps
+from cozmo.lights import Color, Light
+try:
+    from PIL import Image, ImageDraw, ImageColor
+except ImportError:
+    sys.exit('Cannot import from PIL: Do `pip3 install --user Pillow` to install')
+
+''' Color utilities:
+        color_dict (dict): map of color names to regions in color space.
+            Instead of defining a color as a single (R, G, B) point, 
+            colors are defined with a minimum and maximum value for R, G, and B.
+            For example, a point with (R, G, B) = (40, 65, 195) falls in the 'blue' region, 
+            because 0 < 40 < 70, 0 < 65 < 70, and 200 < 195 < 255.
+        
+        map_color_to_light (dict): maps each color name with its cozmo.lights.Light value.
+            Red, green, and blue lights are already defined as constants in lights.py, 
+            but we need to define our own custom Light for yellow.
+
+        color_distance_sqr (fn (color, color_range) => int): returns the squared 
+            euclidean distance of color to color_range. 
+'''
+
+color_dict = {
+'red' : (200, 255, 0, 70, 0, 70), 
+'green' : (0, 70, 200, 255, 0, 70), 
+'blue' : (0, 70, 0, 70, 200, 255), 
+'white' : (200, 255, 200, 255, 200, 255), 
+'black' : (0, 30, 0, 30, 0, 30),
+'yellow' : (200, 255, 200, 255, 0, 70), 
+}
+
+map_color_to_light = {
+'green' : cozmo.lights.green_light, 
+'yellow' : Light(Color(name='yellow', rgb = (255, 255, 0))), 
+'blue' : cozmo.lights.blue_light, 
+'red' : cozmo.lights.red_light
+}
+
+def color_distance_sqr(color, color_range):
+    r, g, b = color
+    minR, maxR, minG, maxG, minB, maxB = color_range
+    rdist_sq = 0
+    gdist_sq = 0
+    bdist_sq = 0
+    if r < minR:
+        rdist_sq = (minR-r)**2
+    if r > maxR:
+        rdist_sq = (maxR-r)**2
+    if g < minG:
+        gdist_sq = (minG-g)**2
+    if g > maxG:
+        gdist_sq = (maxG-g)**2
+    if b < minB:
+        bdist_sq = (minB-b)**2
+    if b > maxB:
+        bdist_sq = (maxB-b)**2
+    return rdist_sq+gdist_sq+bdist_sq
+
+def gray_world(img):
+    ''' Simple color-balancing algorithm for more accurate color distinction.
+
+        Args:
+            nimg: image array containing the RGB data of a PIL image
+    '''
+    nimg = from_pil(img)
+    nimg = nimg.transpose(2, 0, 1).astype(numpy.uint32)
+    mu_g = numpy.average(nimg[1])
+    nimg[0] = numpy.minimum(nimg[0] * (mu_g / numpy.average(nimg[0])), 255)
+    nimg[2] = numpy.minimum(nimg[2] * (mu_g / numpy.average(nimg[2])), 255)
+    return to_pil(nimg.transpose(1, 2, 0).astype(numpy.uint8))
+
+def from_pil(pimg):
+    ''' Converts PIL image into image array for use by gray_world.'''
+    pimg = pimg.convert(mode='RGB')
+    nimg = numpy.asarray(pimg)
```

Comment:
Could rename nimg and pimg to more descriptive var names like image_array and pil_image.